core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -710,13 +710,17 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    * Returns an effect that, if evaluated, will return the lazily computed result
    * of this effect.
    */
-  final def memoize: URIO[R, IO[E, A]] =
-    for {
-      r <- ZIO.environment[R]
-      p <- Promise.make[E, A]
-      l <- Promise.make[Nothing, Unit]
-      _ <- (l.await *> ((self provide r) to p)).fork
-    } yield l.succeed(()) *> p.await
+  final def memoize: URIO[R, IO[E, A]] = {
+    import internal.OneShot
+    (ZIO.environment[R] <*> UIO.effectTotal(OneShot.make[A])).map { case (r, x) =>
+      IO.effectTotal {
+        if (x.isSet)
+          UIO.effectTotal(x.get)
+        else
+          self.tap(a => UIO.effectTotal(x.set(a))).provide(r)
```

Comment:
`OneShot#set` throws if called after the value has been set, but we don't have anything guaranteeing it won't be called twice here, if I'm not mistaken.

Reply:
You mean in sth like `action.memoize.flatMap { a => a <&> a }`? Yeah 🤦‍♂

Reply:
Maybe with a `Ref[Either[A, Promise]]` state machine?